File: CPP/ByTier/E/2024_E_C_K_S.cpp
// Problem: CF 2024 E - C+K+S
// https://codeforces.com/contest/2024/problem/E

/*
Purpose: This code determines whether it's possible to connect two strongly connected directed graphs with n vertices each, such that:
1. Each added edge connects vertices from different graphs
2. Each outgoing vertex has exactly one outgoing edge
3. Each incoming vertex has exactly one incoming edge
4. All resulting cycles have lengths divisible by k

Algorithms/Techniques:
- DFS to compute distances from a root vertex in each graph
- Modular arithmetic to classify vertices by their distance mod k
- String matching (Z-algorithm) to check if a valid assignment exists
- Rolling array technique to handle cyclic permutations

Time Complexity: O(n + m1 + m2) where n is the number of vertices, m1 and m2 are the number of edges in first and second graph respectively.
Space Complexity: O(n + m1 + m2) for storing graphs and auxiliary arrays.

*/
#include <stdlib.h>
#include <algorithm>
#include <iostream>
#include <vector>
#include <utility>

#define vc vector
#define pip pair<int, int>
#define pb push_back
#define ll long long
#define mp(a, b) make_pair(a, b)
#define all(s) s.begin(), s.end()
const int inf = 1e9 + 7;
const ll INF = 1e18;
using namespace std;
vc<vc<int>> g;
void dfs(int v, vc<int> &t) {
  for (int i = 0; i < g[v].size(); i++) {
    int u = g[v][i];
    if (t[u])
      continue;
    t[u] = t[v] + 1;
    dfs(u, t);
  }
}
void roll(vc<int> &arr) {
  int n = arr.size();
  for (int i = n - 1; i > 0; i--)
    swap(arr[i], arr[i - 1]);
}
vc<int> f(vc<int> &a, vc<int> &b) {
  int n = a.size();
  vc<int> s;
  for (int i = 0; i < n; i++)
    s.pb(b[i]);
  for (int i = 0; i < 2 * n; i++)
    s.pb(a[i % n]);
  n *= 3;
  vc<int> h(n);
  int l = 0, r = -1;
  for (int i = 1; i < n; i++) {
    if (i <= r)
      h[i] = min(r - i + 1, h[i - l]);
    while (i + h[i] < n && s[i + h[i]] == s[h[i]])
      h[i]++;
    if (r < i + h[i] - 1)
      r = i + h[i] - 1, l = i;
  }
  return h;
}
void solve() {
  int n, k, m;
  cin >> n >> k;
  g.resize(n);
  vc<int> t1(n), t2(n);
  for (int i = 0; i < n; i++)
    g[i].resize(0);
  vc<int> a(n), b(n);
  for (int i = 0; i < n; i++)
    cin >> a[i];
  cin >> m;
  while (m--) {
    int v, u;
    cin >> v >> u;
    v--, u--;
    g[v].pb(u);
  }
  dfs(0, t1);
  for (int i = 0; i < n; i++)
    g[i].resize(0), cin >> b[i];
  cin >> m;
  while (m--) {
    int v, u;
    cin >> v >> u;
    v--, u--;
    g[v].pb(u);
  }
  dfs(0, t2);
  for (int i = 0; i < n; i++)
    t1[i] %= k, t2[i] %= k;
  int res = 0;
  vc<int> q1(k), q0(k), w1(k), w0(k);
  for (int i = 0; i < n; i++) {
    if (a[i])
      q1[t1[i]]++;
    else
      q0[t1[i]]++;
    if (b[i])
      w1[t2[i]]++;
    else
      w0[t2[i]]++;
  }
  for (int i = 0; i < n; i++)
    res += a[i] - b[i];
  if (abs(res) == n) {
    cout << "YES\n";
    return;
  }
  roll(q1);
  roll(w1);
  vc<int> ans1 = f(q1, w0), ans2 = f(q0, w1);
  for (int i = k; i < 3 * k; i++) {
    if (min(ans1[i], ans2[i]) >= k) {
      cout << "YES\n";
      return;
    }
  }
  cout << "NO\n";
}
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  cout.tie(nullptr);
  int test = 1;
  cin >> test;
  while (test--) {
    solve();
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/2024/2024_E_C_K_S.py
# Problem: CF 2024 E - C+K+S
# https://codeforces.com/contest/2024/problem/E

"""
E. C+K+S

This problem involves determining whether it's possible to construct a bipartite matching between two strongly connected directed graphs such that:
1. Each outgoing vertex in the first graph has exactly one outgoing edge to the second graph.
2. Each incoming vertex in the second graph has exactly one incoming edge from the first graph.
3. The resulting graph maintains cycle lengths divisible by k.

Key techniques used:
- Graph labeling using BFS to assign labels modulo k to vertices based on their connectivity
- KMP string matching algorithm for pattern matching of label sequences
- Set intersection to find valid label alignments

Time Complexity: O(n + m1 + m2) per test case, where n is number of vertices and m1, m2 are edges in each graph.
Space Complexity: O(n + k) for storing labels and KMP tables.
"""

import collections
import sys


def kmp_match(pattern, text):
    if not pattern:
        return []
    pi = [0] * len(pattern)
    k_val = 0
    for q in range(1, len(pattern)):
        while k_val > 0 and pattern[k_val] != pattern[q]:
            k_val = pi[k_val - 1]
        if pattern[k_val] == pattern[q]:
            k_val += 1
        else:
            k_val = 0
        pi[q] = k_val
    q = 0
    occurrences = []
    n_pattern = len(pattern)
    for i in range(len(text)):
        while q > 0 and pattern[q] != text[i]:
            q = pi[q - 1]
        if pattern[q] == text[i]:
            q += 1
        else:
            q = 0
        if q == n_pattern:
            occurrences.append(i - n_pattern + 1)
            q = pi[q - 1] if q > 0 else 0
    return occurrences


def main():
    data = sys.stdin.read().split()
    t = int(data[0])
    index = 1
    output_lines = []
    for _ in range(t):
        n = int(data[index])
        k = int(data[index + 1])
        index += 2
        a = list(map(int, data[index : index + n]))
        index += n
        m1 = int(data[index])
        index += 1
        graph1 = [[] for _ in range(n)]
        for i in range(m1):
            u = int(data[index])
            v = int(data[index + 1])
            index += 2
            graph1[u - 1].append(v - 1)
        b = list(map(int, data[index : index + n]))
        index += n
        m2 = int(data[index])
        index += 1
        graph2 = [[] for _ in range(n)]
        for i in range(m2):
            u = int(data[index])
            v = int(data[index + 1])
            index += 2
            graph2[u - 1].append(v - 1)
        countA = sum(1 for x in a if x == 1)
        countB_prime = sum(1 for x in b if x == 0)
        countB = sum(1 for x in b if x == 1)
        countA_prime = n - countA
        # Check basic compatibility of types
        if countA != countB_prime or countB != countA_prime:
            output_lines.append("NO")
            continue
        label1 = [-1] * n
        label1[0] = 0
        queue = collections.deque([0])
        while queue: # BFS to assign labels based on cycle constraint modulo k
            u = queue.popleft()
            for v in graph1[u]:
                if label1[v] == -1:
                    label1[v] = (label1[u] + 1) % k
                    queue.append(v)
        label2 = [-1] * n
        label2[0] = 0
        queue = collections.deque([0])
        while queue: # BFS to assign labels based on cycle constraint modulo k
            u = queue.popleft()
            for v in graph2[u]:
                if label2[v] == -1:
                    label2[v] = (label2[u] + 1) % k
                    queue.append(v)
        if countA == 0 or countB == 0:
            output_lines.append("YES")
            continue
        # Prepare sequences of labels for matching based on types
        U1 = []
        for i in range(n):
            if a[i] == 1:
                U1.append((label1[i] + 1) % k)
        T2 = []
        for i in range(n):
            if b[i] == 0:
                T2.append(label2[i] % k)
        U2 = []
        for i in range(n):
            if b[i] == 1:
                U2.append((label2[i] + 1) % k)
        T1 = []
        for i in range(n):
            if a[i] == 0:
                T1.append(label1[i] % k)
        A1 = [0] * k
        for x in U1: # Create frequency array of labels from outgoing vertices of graph1
            A1[x] += 1
        A2 = [0] * k
        for x in T2: # Create frequency array of labels from incoming vertices of graph2
            A2[x] += 1
        A3 = [0] * k
        for x in U2: # Create frequency array of labels from outgoing vertices of graph2
            A3[x] += 1
        A4 = [0] * k
        for x in T1: # Create frequency array of labels from incoming vertices of graph1
            A4[x] += 1
        text1 = A1 + A1
        d1_list = kmp_match(A2, text1)
        d1_set = set(d for d in d1_list if 0 <= d < k) # Valid alignment shifts for matching sequence 1
        text2 = A3 + A3
        d2_list = kmp_match(A4, text2)
        d2_set = set(d for d in d2_list if 0 <= d < k) # Valid alignment shifts for matching sequence 2
        found = False
        for d in d1_set:
            d2_val = (k - d) % k
            if d2_val in d2_set:
                found = True
                break
        output_lines.append("YES" if found else "NO")
    sys.stdout.write("\n".join(output_lines))


if __name__ == "__main__":
    main()


# https://github.com/VaHiX/codeForces/